### ISSUE TITLE
Obtain Psets with PMIx_Get and key PMIX_PSET_NAMES

### ISSUE THREAD

**Issue by sonjahapp** [openpmix/openpmix#3140]
### What version of the PMIx Reference Library are you using? 
4.2.5

### Describe how PMIx was installed 
from source via git clone

### Please describe the system on which you are running

* Operating system/version: Fedora 36
* Computer hardware: x86
* Network type: ethernet

-----------------------------

## Details of the problem

Hello!

I would like to clarify the expected behavior of `PMIx_Get` for the key `PMIX_PSET_NAMES`. According to the PMIx standard 5.0, section 14.1.2 (or PMIx 4.1 Section 13.1.2):

> [...], a process can request (via `PMIx_Get`) the process sets to which a given process (including itself) belongs: `PMIX_PSET_NAMES` `"pmix.pset.nms"` `(pmix_data_array_t*)` Returns an array of `char*` string names of the process sets in which the given process is a member.

However, if I try `PMIx_Get(myself, PMIX_PSET_NAMES, NULL, 0, &value)` with OpenPMIx 4.2.5, I get an error `PMIX_ERR_NOT_FOUND` - even in situations where I know that a process is member of at least one PMIx Pset and the PMIx event notification mechanism reported the definition of such a Pset to the process (via `PMIX_PROCESS_SET_DEFINE` event) before.
The PMIx Pset is defined in the PMIx Host using `PMIx_server_define_process_set`. The PMIx Host used is ParTec's ParaStation Management process manager.

Should OpenPMIx take care of returning the correct value for `PMIX_PSET_NAMES`? Or is there some responsibility in the PMIx Host that I am not aware of?

Thank you for the support!

CC @krempel-pt

**Comment by rhc54** [comment]
Sorry for delay - a tad busy right now. Short answer is "yes", we should return the info. However, it depends on how the pset name was defined and passed to the library. I'd have to check the code to see if we cache the info somewhere that it can be retrieved, or if we rely on the host to cache and return that info.

I know it works correctly when the pset is defined at time of application launch, but it sounds like you are talking about a pset that is dynamically defined after launch - or else the event would not have fired. True?

**Comment by sonjahapp** [comment]
Thank you for your time!

> ... if we cache the info somewhere that it can be retrieved, or if we rely on the host to cache and return that info

As far as I can see, `pmix_server_globals.psets` is updated with any new pset defined dynamically via `PMIx_server_define_process_set` (see [here](https://github.com/openpmix/openpmix/blob/master/src/server/pmix_server.c#L3028)). Perhaps this could help to retrieve the info?

> it sounds like you are talking about a pset that is dynamically defined after launch - or else the event would not have fired. True?

Yes, that is true. In ParaStation Management we add psets (at least one)  dynamically shortly after launch.

**Comment by rhc54** [comment]
I think #3142 should fix it - can you give it a try?

**Comment by sonjahapp** [comment]
Thanks! I tried your patch - without success unfortunately.

I've extended  `examples/pset.c` with the following snippet right before the `done` label:
```C
    pmix_value_t *val;
    if (PMIX_SUCCESS != (rc = PMIx_Get(&myproc, PMIX_PSET_NAMES, NULL, 0, &val))) {
        fprintf(stderr, "[%s:%d] PMIx_Get PMIX_PSET_NAMES returned %s\n", myproc.nspace,
                myproc.rank, PMIx_Error_string(rc));
        goto done;
    }
    fprintf(stderr, "[%s:%d] psets %s\n", myproc.nspace, myproc.rank, val->data.string);
    PMIX_VALUE_RELEASE(val);

done:
```

I still get `PMIX_ERR_NOT_FOUND` for PMIX_PSET_NAMES even though the event mechanism reported a dynamically defined pset before.

**Comment by rhc54** [comment]
Would you mind adding `PMIX_MCA_pmix_client_get_verbose=5` and `PMIX_MCA_pmix_server_get_verbose=5` to your environment and running it again? I don't have an easy way to reproduce this problem. Also, just to verify - the change was made to the PMIx `master` branch, so you'd need to have both ParaStation and your app built against it. I haven't looked at backporting it to the v4.2 series yet.

**Comment by sonjahapp** [comment]
I noticed that your change is on `master`. Unfortunately, ParaStation Management's PMIx plugin is not yet ready for PMIx 5 and does not work with the PMIx `master` at the moment. This makes testing tricky for me as well. :see_no_evil:  

I tried to backport your changes to v4.2.5 with a naive approach - just patch `pmix_server_get.c` with the same changes you made on `master` . For a `PMIx_Get` on `PMIX_PSET_NAMES` I see the following error:
```
PMIX ERROR: PMIX_ERR_UNPACK_FAILURE in file ../../../../../src/mca/gds/hash/gds_hash.c at line 1575
```
which results in reporting `PMIX_ERR_NOT_FOUND` on the client side.

I will try with more debug output as you suggest.

**Comment by rhc54** [comment]
FWIW: I'm just finishing a new `pctrl` tool for issuing `PMIx_Job_control` requests, one of which is to "define new pset". Should let me debug the patch.

**Comment by rhc54** [comment]
Okay, I can now locally reproduce the problem - I'll try to debug it.

**Comment by rhc54** [comment]
I tracked down the problem - the client was expecting the data returned by server_get to be in a different format than the updated code I provided was sending. I included the new tool in the patch as having a cmd line way of requesting job control operations seemed useful.

**Comment by rhc54** [comment]
I'll look at the v4.2 backport after #3142 is committed.

**Comment by sonjahapp** [comment]
Thank you very much for tracking down the problem, Ralph! 

I have patched v4.2.5 locally with the changes you made in #3142 and I can confirm that I obtain the expected psets with `PMIx_Get` on `PMIX_PSET_NAMES` now! :+1:

**Comment by rhc54** [comment]
> We have not yet looked deeper into it on our side. But if it _should_ actually not be a problem to switch from v4.2 to v5 (at least not for interfaces such as `PMIx_Init`), shall I open another issue to follow-up on this?

I'm not sure there is much I can do to help, other than perhaps providing advice. The code in that area hasn't changed in quite some time, and we test (at a per-PR level) not only that v5 server to v5 client works, but also that v5 server to any client version works and that v5 client works with any server version.

So we aren't seeing any compatibility issues, nor (other than this) are we hearing of any out there. I can only suspect that something isn't quite right in ParaStation? Or maybe something about how you are testing it? 🤷‍♂️ 

You are welcome to open an issue if you want to ask questions to help chase down the problem.
